### ISSUE TITLE
pytest doesn't run tests and runs coverage on virtualenv sources

### ISSUE THREAD

**Issue by dodumosu** [Pylons/pyramid#2812]
Hi, new to Pyramid and following the quick tutorial. Got to the testing section and I had issues getting pytest to execute the tests with the commands in the documentation (likely the way pytest finds tests has changed).

[this](http://imgur.com/a/1bASs) links to a screenshot that shows the issue.

I was able to get the tests to run using `py.test hello_world/tests.py --cov --cov-report=term-missing`, but the coverage results weren't limited to the sources in the `hello_world` folder.

**Comment by stevepiercy** [comment]
Traceback text > screenshots.

Did you follow all the steps in all previous steps? The Quick Tutorial may not work if you jump into a step without following pre-requisite steps.

Based on your path, there are no tests for http://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tutorial/hello_world.html

That indicates you did not follow the steps in order.

**Comment by dodumosu** [comment]
Steps to recreate:

- using `pcreate -s pyramid_jinja2_starter hello_world`, create new project
- `cd hello_world`
- created project-local virtualenv using `pyvenv venv/3` (intended to keep one for Py3, and one for Py2)
- activated project-local virtualenv using `source venv/3/bin/activate`
- installed project to virtualenv using `pip install -e .`. this used `setup.py` in the project folder
- tested using `pserve development.ini`. waitress ran, access wasn't logged to screen
- edited `setup.py` to contain:
```
    tests_require = [
        'WebTest >= 1.3.1',
        'pytest',
        'pytest-cov'
    ]

    # ...
    extras_require={'testing': tests_require},
```
- installed test dependencies using `pip install -e ".[testing]". this installed WebTest 2.0.23, beautifulsoup4-4.5.1, coverage-4.2, py-1.4.1, pytest-3.0.4, pytest-cov-2.4.0, six-1.10.0
- ran `py.test --cov --cov-report=term-missing`. this found tests in the local virtualenv. partial output below:
```
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.5.2, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /Users/dipoodumosu/work/code/python/learning-pyramid/hello_world, inifile:
plugins: cov-2.4.0
collected 4421 items

venv/3/lib/python3.5/site-packages/bs4/tests/test_builder_registry.py ...........
venv/3/lib/python3.5/site-packages/bs4/tests/test_htmlparser.py ..................................................
venv/3/lib/python3.5/site-packages/bs4/tests/test_soup.py ...............................................
venv/3/lib/python3.5/site-packages/bs4/tests/test_tree.py .............................................................................................................................................................................................................................
venv/3/lib/python3.5/site-packages/pyramid/tests/test_asset.py ...........
venv/3/lib/python3.5/site-packages/pyramid/tests/test_authentication.py .....................................................................................................................................................
venv/3/lib/python3.5/site-packages/pyramid/tests/test_authorization.py .............
venv/3/lib/python3.5/site-packages/pyramid/tests/test_compat.py .....
venv/3/lib/python3.5/site-packages/pyramid/tests/test_decorator.py ...
venv/3/lib/python3.5/site-packages/pyramid/tests/test_encode.py ..............
venv/3/lib/python3.5/site-packages/pyramid/tests/test_events.py ......................................
venv/3/lib/python3.5/site-packages/pyramid/tests/test_exceptions.py ..........
venv/3/lib/python3.5/site-packages/pyramid/tests/test_httpexceptions.py ........................................................
venv/3/lib/python3.5/site-packages/pyramid/tests/test_i18n.py .............................................
venv/3/lib/python3.5/site-packages/pyramid/tests/test_integration.py
```
- cd-ed into the internal hello_world folder
- reran `py.test --cov --coverage-report=term-missing`. output is below:
```
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.5.2, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /Users/dipoodumosu/work/code/python/learning-pyramid/hello_world, inifile:
plugins: cov-2.4.0
collected 0 items


---------- coverage: platform darwin, python 3.5.2-final-0 -----------
Name                                                                                                                                      Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/_argcomplete.py                  33     32     3%   2-98, 101
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/_code/code.py                   570    568     1%   1-833, 837-859
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/assertion/__init__.py            73     69     5%   4-91, 101-162
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/assertion/rewrite.py            561    530     6%   3-58, 61, 76-77, 82-85, 96-145, 149-150, 153-156, 166-167, 171-172, 176-940
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/cacheprovider.py                139    121    13%   8-15, 20-92, 97, 101, 103-118, 122-129, 131-176, 178-183, 189-205, 207-245
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/capture.py                      328    300     9%   5-52, 55-87, 97-103, 106-108, 111-118, 122-249, 253-281, 287-311, 316-358, 363-368, 371-380, 384-422, 425-490
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/compat.py                       122    111     9%   4-73, 78, 82-83, 85-90, 97, 102-230
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/config.py                       841    713    15%   2-187, 195, 204-214, 217-225, 232-236, 240, 250-262, 277-285, 289, 295-297, 299, 312, 315-316, 321, 324-327, 330-486, 492, 508, 514-633, 636, 653-728, 730, 738, 742-748, 752-857, 862-926, 930-1050, 1058, 1069, 1072-1073, 1082, 1084-1100, 1105-1114, 1121-1125, 1128, 1141, 1143-1151, 1155-1280, 1287, 1289, 1293-1308
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/debugging.py                     88     81     8%   2-22, 24-31, 35-37, 43-122
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/doctest.py                      174    165     5%   2-48, 52, 54-57, 59, 63, 67-331
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/fixtures.py                     656    594     9%   1-19, 23-120, 125-163, 168-170, 173, 176-611, 617-726, 746-946, 956-974, 985-987, 990-1054, 1058, 1062-1063, 1065, 1076-1081, 1085, 1088, 1099, 1108, 1111-1131
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/helpconfig.py                    96     88     8%   2-31, 34-52, 54-61, 63-123, 126-131, 134-141
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/junitxml.py                     228    226     1%   11-226, 230-409
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/main.py                         525    380    28%   2-81, 85, 97-124, 127, 133-138, 141, 144, 148, 151-154, 157, 163, 166, 173, 180-192, 196-199, 204-238, 263, 268-302, 311-314, 317-323, 327, 329, 332-337, 341-458, 462-471, 475-478, 482-535, 547, 550, 553-555, 558-565, 568, 579, 587, 598, 615-620, 622, 626, 629, 636-640, 644, 652-658, 662, 665, 668, 671, 676-698, 702, 706-710, 714-765
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/mark.py                         168    160     5%   2-45, 48-60, 68-162, 165-326
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/monkeypatch.py                  125    113    10%   3-103, 109, 127-132, 136, 140, 144-258
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/nose.py                          42     41     2%   3-52, 54-71
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/pastebin.py                      61     59     3%   2-14, 18-92
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/python.py                       851    841     1%   3-88, 90-91, 93-112, 128-157, 163, 166-1578
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/resultlog.py                     86     84     2%   5-14, 18-107
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/runner.py                       365    334     8%   2-53, 55-142, 148-157, 164-168, 171-242, 248-332, 338-349, 356-357, 366-382, 386-458, 464, 472-570
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/setuponly.py                     46     45     2%   1-69, 72
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/setupplan.py                     12     11     8%   1-19, 22-23
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/skipping.py                     251    247     2%   2-24, 26-31, 56-370
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/terminal.py                     452    359    21%   5-47, 52-57, 62-63, 68-69, 72-92, 104, 112-134, 136-142, 144, 148, 152, 156-175, 177-232, 234-236, 238, 240, 247, 249, 259, 261, 267-269, 272, 276, 281-282, 287, 295, 298, 308, 310-316, 318-539, 547-581
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/tmpdir.py                        62     55    11%   2-15, 19-88, 102-123
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py     429    293    32%   67-163, 170-176, 179-197, 199-200, 206-218, 222, 225-228, 232-238, 246-247, 251, 257-263, 266-273, 277-332, 336, 341, 348-350, 376, 380-381, 384, 389-458, 462-472, 474, 480-487, 496-520, 525-592, 598, 607-610, 623-630, 643-647, 650-653, 659-660, 663-664, 668, 671-672, 676-687, 693-696, 699-710, 713-723, 740-743, 747, 752-765, 770-774, 783-799
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/pkg_resources/__init__.py              1499   1497     1%   17-2456, 2460-2956
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_apipkg.py                           138    119    14%   8-47, 50, 58-110, 115-116, 122, 129-130, 133-181
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_builtin.py                          188    187     1%   1-134, 137-248
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_code/code.py                        523    521     1%   1-772, 776-786
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_error.py                             50     47     6%   5-60, 65-88
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_io/terminalwriter.py                234    151    35%   16-23, 37-41, 47, 51, 58-108, 112, 114, 129-132, 135, 137, 154, 169, 181, 186, 193, 206-211, 216, 220-246, 250-251, 254-256, 259, 263-324, 334-347
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_path/common.py                      244    186    24%   3-12, 15-41, 44-47, 53-59, 61, 68-95, 98-110, 113-171, 189, 192, 211-267, 270, 286-314, 335, 337-347, 349, 351, 353, 361, 364-365, 370-372, 381-382, 385, 394, 399-401
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_path/local.py                       628    549    13%   4-16, 19-138, 154, 157-160, 163, 167-171, 174-177, 180-277, 286, 288, 296, 300, 304, 307, 313, 315-317, 332, 337-338, 344-355, 360-363, 373-374, 383-386, 397-579, 583-911
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_std.py                               11      5    55%   12-16
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/py/_xmlgen.py                           170    168     1%   7-23, 26-253
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/pytest_cov/compat.py                     19     15    21%   1-18, 23-25, 29-31
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/pytest_cov/engine.py                    162    151     7%   3-31, 36, 39, 44, 51-145, 150-291
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/venv/3/lib/python3.5/site-packages/pytest_cov/plugin.py                    158    134    15%   2-17, 25-26, 31-44, 49-99, 103-177, 180, 185-186, 190-192, 194-221, 226, 234-291
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                                                     11408  10350     9%

=========================================================================== no tests ran in 1.57 seconds ============================================================================
```

output of `ls -al` in that folder:
```
total 56
drwxr-xr-x  11 dipoodumosu  staff    374 Nov 13 02:21 .
drwxr-xr-x  13 dipoodumosu  staff    442 Nov 13 02:20 ..
-rw-r--r--   1 dipoodumosu  staff  10393 Nov 13 02:21 .coverage
-rw-r--r--   1 dipoodumosu  staff    766 Nov 13 01:55 __init__.py
drwxr-xr-x   5 dipoodumosu  staff    170 Nov 13 02:01 __pycache__
drwxr-xr-x   5 dipoodumosu  staff    170 Nov 13 01:55 locale
-rw-r--r--   1 dipoodumosu  staff     97 Nov 13 01:55 resources.py
drwxr-xr-x   6 dipoodumosu  staff    204 Nov 13 01:55 static
drwxr-xr-x   3 dipoodumosu  staff    102 Nov 13 01:55 templates
-rw-r--r--   1 dipoodumosu  staff    496 Nov 13 01:55 tests.py
-rw-r--r--   1 dipoodumosu  staff    157 Nov 13 01:55 views.py
```

**Comment by dodumosu** [comment]
Using a non-local virtualenv (I deleted the tree at `hello_world/venv` and created an external virtualenv) only changed the behaviour in that pytest didn't find any tests in the project folder.

**Comment by stevepiercy** [comment]
Which step in the Quick Tutorial are you trying to get to work? Please provide the URL.

We do not use `activate` in any of the docs for Pyramid. We use either the path to the virtual environment, or an environment variable set to the virtual environment's path, when invoking commands.

I don't know why pytest collected 4421 items. It should be collecting only a few, if you followed the Quick Tutorial instructions and don't go off-roading with `activate` and anything else not prescribed.

**Comment by dodumosu** [comment]
@stevepiercy sorry to seem like a tool. I understand you're giving your time to this, and I really don't want to stress anyone out. That said, the two pain points are:
- py.test doesn't seem to find the tests for the project unless I manually pass it the test module
- coverage returns coverage for the sources in the py.test package in the virtualenv, not the project.

here's the output from my last run (with the full path to the virtualenv executable specified):
```
(pyramid-quick-tour) ➜  hello_world git:(develop) ✗  deactivate
➜  hello_world git:(develop) ✗  ~/.virtualenvs/pyramid-quick-tour/bin/py.test --cov --cov-report=term-missing
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.5.2, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /Users/dipoodumosu/work/code/python/learning-pyramid/hello_world, inifile:
plugins: cov-2.4.0
collected 0 items


---------- coverage: platform darwin, python 3.5.2-final-0 -----------
Name                                                                                                                 Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------------------------------------------------------------------------
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/_argcomplete.py                  33     32     3%   2-98, 101
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/_code/code.py                   570    568     1%   1-833, 837-859
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/assertion/__init__.py            73     69     5%   4-91, 101-162
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/assertion/rewrite.py            561    530     6%   3-58, 61, 76-77, 82-85, 96-145, 149-150, 153-156, 166-167, 171-172, 176-940
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/cacheprovider.py                139    121    13%   8-15, 20-92, 97, 101, 103-118, 122-129, 131-176, 178-183, 189-205, 207-245
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/capture.py                      328    300     9%   5-52, 55-87, 97-103, 106-108, 111-118, 122-249, 253-281, 287-311, 316-358, 363-368, 371-380, 384-422, 425-490
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/compat.py                       122    111     9%   4-73, 78, 82-83, 85-90, 97, 102-230
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/config.py                       841    713    15%   2-187, 195, 204-214, 217-225, 232-236, 240, 250-262, 277-285, 289, 295-297, 299, 312, 315-316, 321, 324-327, 330-486, 492, 508, 514-633, 636, 653-728, 730, 738, 742-748, 752-857, 862-926, 930-1050, 1058, 1069, 1072-1073, 1082, 1084-1100, 1105-1114, 1121-1125, 1128, 1141, 1143-1151, 1155-1280, 1287, 1289, 1293-1308
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/debugging.py                     88     81     8%   2-22, 24-31, 35-37, 43-122
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/doctest.py                      174    165     5%   2-48, 52, 54-57, 59, 63, 67-331
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/fixtures.py                     656    594     9%   1-19, 23-120, 125-163, 168-170, 173, 176-611, 617-726, 746-946, 956-974, 985-987, 990-1054, 1058, 1062-1063, 1065, 1076-1081, 1085, 1088, 1099, 1108, 1111-1131
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/helpconfig.py                    96     88     8%   2-31, 34-52, 54-61, 63-123, 126-131, 134-141
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/junitxml.py                     228    226     1%   11-226, 230-409
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/main.py                         525    380    28%   2-81, 85, 97-124, 127, 133-138, 141, 144, 148, 151-154, 157, 163, 166, 173, 180-192, 196-199, 204-238, 263, 268-302, 311-314, 317-323, 327, 329, 332-337, 341-458, 462-471, 475-478, 482-535, 547, 550, 553-555, 558-565, 568, 579, 587, 598, 615-620, 622, 626, 629, 636-640, 644, 652-658, 662, 665, 668, 671, 676-698, 702, 706-710, 714-765
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/mark.py                         168    160     5%   2-45, 48-60, 68-162, 165-326
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/monkeypatch.py                  125    113    10%   3-103, 109, 127-132, 136, 140, 144-258
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/nose.py                          42     41     2%   3-52, 54-71
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/pastebin.py                      61     59     3%   2-14, 18-92
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/python.py                       851    841     1%   3-88, 90-91, 93-112, 128-157, 163, 166-1578
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/resultlog.py                     86     84     2%   5-14, 18-107
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/runner.py                       365    334     8%   2-53, 55-142, 148-157, 164-168, 171-242, 248-332, 338-349, 356-357, 366-382, 386-458, 464, 472-570
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/setuponly.py                     46     45     2%   1-69, 72
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/setupplan.py                     12     11     8%   1-19, 22-23
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/skipping.py                     251    247     2%   2-24, 26-31, 56-370
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/terminal.py                     452    359    21%   5-47, 52-57, 62-63, 68-69, 72-92, 104, 112-134, 136-142, 144, 148, 152, 156-175, 177-232, 234-236, 238, 240, 247, 249, 259, 261, 267-269, 272, 276, 281-282, 287, 295, 298, 308, 310-316, 318-539, 547-581
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/tmpdir.py                        62     55    11%   2-15, 19-88, 102-123
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py     429    293    32%   67-163, 170-176, 179-197, 199-200, 206-218, 222, 225-228, 232-238, 246-247, 251, 257-263, 266-273, 277-332, 336, 341, 348-350, 376, 380-381, 384, 389-458, 462-472, 474, 480-487, 496-520, 525-592, 598, 607-610, 623-630, 643-647, 650-653, 659-660, 663-664, 668, 671-672, 676-687, 693-696, 699-710, 713-723, 740-743, 747, 752-765, 770-774, 783-799
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/pkg_resources/__init__.py              1508   1506     1%   2-2518, 2522-3051
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_apipkg.py                           138    119    14%   8-47, 50, 58-110, 115-116, 122, 129-130, 133-181
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_builtin.py                          188    187     1%   1-134, 137-248
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_code/code.py                        523    521     1%   1-772, 776-786
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_error.py                             50     47     6%   5-60, 65-88
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_io/terminalwriter.py                234    151    35%   16-23, 37-41, 47, 51, 58-108, 112, 114, 129-132, 135, 137, 154, 169, 181, 186, 193, 206-211, 216, 220-246, 250-251, 254-256, 259, 263-324, 334-347
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_path/common.py                      244    186    24%   3-12, 15-41, 44-47, 53-59, 61, 68-95, 98-110, 113-171, 189, 192, 211-267, 270, 286-314, 335, 337-347, 349, 351, 353, 361, 364-365, 370-372, 381-382, 385, 394, 399-401
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_path/local.py                       628    549    13%   4-16, 19-138, 154, 157-160, 163, 167-171, 174-177, 180-277, 286, 288, 296, 300, 304, 307, 313, 315-317, 332, 337-338, 344-355, 360-363, 373-374, 383-386, 397-579, 583-911
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_std.py                               11      5    55%   12-16
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/py/_xmlgen.py                           170    168     1%   7-23, 26-253
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/pytest_cov/compat.py                     19     15    21%   1-18, 23-25, 29-31
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/pytest_cov/engine.py                    162    151     7%   3-31, 36, 39, 44, 51-145, 150-291
/Users/dipoodumosu/.virtualenvs/pyramid-quick-tour/lib/python3.5/site-packages/pytest_cov/plugin.py                    158    134    15%   2-17, 25-26, 31-44, 49-99, 103-177, 180, 185-186, 190-192, 194-221, 226, 234-291
--------------------------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                                11417  10359     9%

=========================================================================== no tests ran in 1.72 seconds ============================================================================
➜  hello_world git:(develop) ✗  pwd
/Users/dipoodumosu/work/code/python/learning-pyramid/hello_world/hello_world
➜  hello_world git:(develop) ✗  cd ..
➜  hello_world git:(develop) ✗  ls
CHANGES.txt            README.txt             hello_world            message-extraction.ini setup.py
MANIFEST.in            development.ini        hello_world.egg-info   setup.cfg
➜  hello_world git:(develop) ✗  ls hello_world
__init__.py  __pycache__  locale       resources.py static       templates    tests.py     views.py
➜  hello_world git:(develop) ✗
```

**Comment by stevepiercy** [comment]
I can't help you unless and until you provide the URL of the step in the Quick Tutorial you are trying to get to work. Do that as a first step, so I have a point of reference.

I don't know what you mean by "testing section" because there are two Steps that reference testing and multiple steps that tell the user to run tests.

**Comment by dodumosu** [comment]
[here it is](http://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tutorial/unit_testing.html).

**Comment by stevepiercy** [comment]
Looking back at your earlier posts, we don't use `pcreate -s pyramid_jinja2_starter hello_world` anywhere in Quick Tutorial. This indicates you went off-road and over the cliff. I have no idea what tutorial you followed now.

FWIW, when I follow the steps in the tutorial in order, including [Requirements](http://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tutorial/requirements.html), without deviation, I have no problem running the tests.

`activate` could be one issue. `activate` is not "wrong", but we don't support it because it causes confusion about what is getting loaded from where. Using an explicit path to the virtual environment avoids this confusion altogether from a tutorial perspective.

I suggest that you start from the beginning of the Quick Tutorial and follow the Steps in order. Do not deviate from the path laid out before you, if you want support.

If there is, in fact, a problem in the Quick Tutorial as it is written and after you follow those steps, then please reopen this issue or open a new issue.

**Comment by dodumosu** [comment]
@stevepiercy i corrected the link, it was the Quick Tour. also, i was adjusting unconsciously while following, because when you get to the [part of the Quick Tour introducing scaffolds](http://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tour.html#quick-project-startup-with-scaffolds), you're supposed to be able to execute `pcreate` from the shell, but unless you installed Pyramid globally, the `pcreate` command needs to be run from the virtualenv.

I followed the steps, no deviation, and the tests still don't run, coverage still reports on the sources in the _pytest, py, and pytest_cov packages in the virtualenv.

**Comment by stevepiercy** [comment]
We use `activate` anywhere in Quick Tour or anywhere in the docs.

We install Pyramid in a virtual environment. We never install anything "globally".

I don't know what you did, but it seems like you've gone far off the beaten path. Start from the beginning. Don't go off-road.

**Comment by dodumosu** [comment]
please see #2813 for what I meant by "global". thanks for all your help.

**Comment by stevepiercy** [comment]
#2813 is a separate matter. Our instructions for [installation in the Quick Tour](http://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tour.html#installation) are terse but sufficient, and provide links for further details about decisions we have made, including [Why use $VENV/bin/pip instead of source bin/activate, then pip](http://docs.pylonsproject.org/projects/pyramid/en/latest/narr/install.html#venv-bin-pip-vs-source-bin-activate).

The matter at hand is that you used `activate`, and we don't provide support for that method. All your invocations of `pip whatever` should have been `$VENV/bin/pip whatever`. Same applies for invoking any binary.

That said, I'll merge your PR #2813. Thanks for catching those two instances where `$VENV/bin/` was omitted.

**Comment by dodumosu** [comment]
hello again. sorry to seem to flog a dead horse, but the issue is not so much with Pyramid as that `pytest` (by default) won't collect and execute tests from a `tests.py` module. that's why I raised this issue, as beginners might not understand why the tests aren't running as advertised in the Quick Tour. these same tests would run using `nose`. the explanation i give below may seem unrelated, but it's not.

let's assume the following project structure, consistent with what you get from creating a project using the pyramid-jinja2 scaffold:
```
- hello_world
|- other-stuff
|- hello_world
|   |- __init__.py
|   |- tests.py
    |- other stuff
```
if you execute `py.test` in the parent `hello_world` folder, it won't find the tests there. if you execute `nose`, on the other hand, it does. if you rename `tests.py` to, say, `test_all.py`, _both_ pytest and nose pick up the tests. i've done a little (not a lot) of research, and the only way i've found so far to change pytest's behaviour is to create a configuration file for it, or directly pass the test module to it as an argument. it doesn't matter if you replace the `tests.py` module with a package, pytest still won't execute the tests in it by default.

**Comment by dodumosu** [comment]
After discussing the issue a bit on the IRC channel, it came to light that the release version of Pyramid-Jinja2's scaffold doesn't contain a `pytest.ini` template, but one was added to the repo on Sept 4. i guess if you installed pyramid-jinja2 from the repository, you wouldn't have this issue.

**Comment by mmerickel** [comment]
The tutorial really was broken because pyramid_jinja2 hadn't been released with the new scaffold. It should be fixed now as I just released 2.7. I'd appreciate @dodumosu confirm that when he has a chance.

**Comment by dodumosu** [comment]
@mmerickel yes, it works. In fact, the scaffold already includes test requirements in `setup.py`, so i created a new PR, #2819, to change the docs to reflect that the user need not edit `setup.py` anymore.
